### ISSUE TITLE
Single File Route

### ISSUE THREAD

**Issue by brillout** [brillout/vite-plugin-ssr#341]
### Description

Being able to define the entire routing of the app in a single file.

**Comment by brillout** [comment]
How about following design.

```bash
/pages/index/index.page.js
/pages/about/index.page.js
/pages/product/index.page.js
/pages/product-list/index.page.js
/routes.js
```

```js
// /routes.js

export default {
  routes: [
    '/pages/index/index.page.js',
    '/pages/about/index.page.js',
    {
      path: '/pages/product/index.page.js',
      route: '/product/@id'
      /* Or a Route Function:
      route: pageContext => {
        // ...
      }
      */
    },
    {
      path: '/pages/product-list/index.page.js',
      route: '/products'
    },
  ],
  // By default all routes need to be listed in `/routes.js` (if `/routes.js` exists).
  // Set to `true` to allow routes to be skipped (we recommend against this).
  allowUnlistedRoutes: false
}
```

```js
function Menu() {
  return <>
    <Link to="index" /> // => <a href="/" />
    <Link to="about" /> // => <a href="/about" />
    <Link to="product" id={42} /> // => <a href="/product/42" />
    <Link to="product-list" /> // => <a href="/products" />
  </>
}
```

Open question: how to make it work with TypeScript? So that TypeScript complains upon `<Link to="unknown-page" />` or `<Link to="product-list" id={"42"} />` (because `"42"` is a `string` but should be a `number`).

I can see negative reactions about this design. Feel free to criticize.

**Comment by redbar0n** [comment]
One option could be to use a tagged template literal string. Which has [typescript support](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html). It would also allow to eliminate the documentation /  comment in the aforementioned example, since it would be more direct.

```jsx
function Menu() {
  return <>
    <Link to=`/` />
    <Link to=`/about` />
    <Link to=`/product/${42}` />
    <Link to=`/products` />
  </>
}
```

**Comment by redbar0n** [comment]
I see there’s some precedence to the API that I suggested: https://solito.dev/usage/link#wrapper-components

**Comment by brillout** [comment]
> Open question: how to make it work with TypeScript? So that TypeScript complains upon `<Link to="unknown-page" />` or `<Link to="product-list" id={"42"} />` (because `"42"` is a `string` but should be a `number`).

Does Solito has such TypeScript support?

>  One option could be to use a tagged template literal string.

So far, I can't see how TypeScript's template literal strings would help.

One aspect of Single Route Files that is interesting is that the user can change the URL without having to modify each link. That's why I'm leaning towards `<Link to="product" id={42} /> `. But I can see that to be a matter of taste and use case. Small apps don't need that.

**Comment by redbar0n** [comment]
> Does Solito has such TypeScript support?

It [doesn't seem](https://solito.dev/usage/link) like it. It simply replicates NextJS' API.

> So far, I can't see how TypeScript's template literal strings would help.

On second glance, I think you're right. TypeScript [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html) that I linked to are a different thing. What I had in mind - type checking inside tagged template literal strings -  is still an open TypeScript issue (not very likely to get implemented, it seems): https://github.com/microsoft/TypeScript/issues/29432

>  change the URL without having to modify each link

I see. I think the general web native way of doing it - and what Rails did - is to simply provide a re-route to the new URL.  That way, outdated links in ones app would work the same as the outdated links to one's page that had already been shared on the web: they would be re-routed to the new URL.

**Comment by redbar0n** [comment]
> Open question: how to make it work with TypeScript? So that TypeScript complains upon <Link to="unknown-page" /> or <Link to="product-list" id={"42"} /> (because "42" is a string but should be a number).

Did some more research into this, since you mentioned it is a blocker.

In addition to [pathpida](https://github.com/aspida/pathpida), check out [next-type-safe-routes](https://github.com/ckastbjerg/next-type-safe-routes) or [nextjs-routes](https://github.com/tatethurston/nextjs-routes) or [next-typed-routes](https://github.com/jagaapple/next-typed-routes) for inspiration.

The API of [next-type-safe-routes](https://github.com/ckastbjerg/next-type-safe-routes) seems the nicest (clearest, most straightforward/intuitive, least boilerplate). The automatic route listing (route autocomplete) is quite nice.

> next-type-safe-routes parses the /pages folder in your Next.js app and generates types for all the pages and API routes in the application. These types can then be used to ensure that you only link to pages (and only use API routes) that actually exists in your application.

> due to the strictness, we can also determine which parameters are needed for dynamic routes.

So, the only issue with using either of those directly is that they all presume filesystem routing (being for NextJS).

### Maybe [typesafe-routes](https://github.com/kruschid/typesafe-routes) what you're looking for?

Actually, it turns out you _can_ use TypeScript [Template Literal Types](https://devblogs.microsoft.com/typescript/announcing-typescript-4-1-beta/#template-literal-types) to get type checking inside tagged template literal strings, like I  had in mind. This library does it, see the GIF animation in that repo.

As long as it is possible to [wrap](https://solito.dev/usage/link) the <Link> component that `vite-plugin-ssr` provides, so that a [Solito.dev](https://solito.dev/) like library could be developed on top, for universal/crossplatform-app support. With the goal of having vite-plugin-ssr (Vite) instead of NextJS (Webpack) as the base for such crossplatform apps.

**Comment by brillout** [comment]
Neat. Indeed, `infer` with template literal types is a solution.

We can even make the route types defined by the user in `route.ts` available globally by using `declare module` (e.g. https://telefunc.com/typescript#getcontext).

I think we can make it work.

Although Route Functions routes cannot be typed, but that's expected.

> As long as it is possible to [wrap](https://solito.dev/usage/link) the component that vite-plugin-ssr provides

Yes, that's the idea. So far, I don't see a problem with that.

**Comment by redbar0n** [comment]
Came across this. Could be potential inspiration for typesafe routes.

> solito’s useParam() hook takes in a parse() function to let you turn the string → number. i’ve found it to be nice for dealing with numbers and validating enums it comes with type safety + inference too: https://solito.dev/usage/params#parsing-values from [Fernando's tweet](https://twitter.com/FernandoTheRojo/status/1567976793285734400?s=20&t=KsKrzTO3OAW7S0DXNnKVSg)

**Comment by redbar0n** [comment]
[react-router-url-params](https://github.com/mcctomsk/react-router-url-params) is another one, for typesafe query parameters. It was made this summer.

**Comment by brillout** [comment]
My idea is that TypeScript for query params just works if we treat param serialization as implementation detail defined by the user.

```jsx
// `order` is typesafe
<Link to="product-search" order={"desc"} />
```

```js
// /routes.js

export default {
  routes: [
    {
      Page: '/pages/product-search.page.js',
      toUrl: ({ order }: { order: 'asc' | 'desc' }) => `/product/search?order={order}`,
      route: '/product/search'
    },
  ]
}
```

**Comment by patryk-smc** [comment]
I'm a big fan of [Tanstack React Location](https://react-location.tanstack.com/) and upcoming [Tanstack Router](https://tanstack.com/router). I think it's API is quite well thought. See this single route file example from Tanstack Router: https://github.com/TanStack/router/blob/572318b1b132307be02857b97ae05b2040363ccf/examples/kitchen-sink/src/index.tsx#L91

I really like how React components can be used instead of file path as proposed in this issue. Also, [pending state feature](https://react-location.tanstack.com/guides/pending-states) sounds incredible.

**Comment by redbar0n** [comment]
The API of React Location looks a lot like the API of (Remix and) React Router v6 : 
https://gist.github.com/redbar0n/f6ec12264ff9d58e243cc516e4e3f41b?permalink_comment_id=4312122#gistcomment-4312122

**Comment by brillout** [comment]
The overall design is starting to crystalize (it also includes better support for nested layouts).

One open question: https://stackoverflow.com/questions/73957822/get-import-path-of-dynamic-import-as-typescript-string-type, in case someone knows this. I don't think it's possible, but who knows with TypeScript 😅.

**Comment by redbar0n** [comment]
blitz-js has another way to achieve typesafe routing which looks quite neat:

https://twitter.com/aleksandrasays/status/1559157742559825920?s=20&t=uaV2ruG_UcOlWxDwGhk9yQ

**Comment by brillout** [comment]
> blitz-js has another way to achieve typesafe routing which looks quite neat:
> 
> https://twitter.com/aleksandrasays/status/1559157742559825920?s=20&t=uaV2ruG_UcOlWxDwGhk9yQ

👍 Neat idea to make the link target a TS object, so you can definition-jump directly to the route file. Neat.

**Comment by brillout** [comment]
@redbar0n Yes I'm also not too fond of their verbosity. While I share the sentiment to keep things succinct, I'm thinking multi-dimensional arrays are a bit too extreme, i.e. not explicit enough. Let's see if the solution I come up with is succinct enough (so far I think so).

**Comment by redbar0n** [comment]
Ok, excited to see how it will look. I wonder how much explicitness is needed if we can have some [Convention over Configuration](https://rubyonrails.org/doctrine#convention-over-configuration).

**Comment by robinelvin** [comment]
Coming from React Router I use their [Javascript Route Definitions](https://reactrouter.com/en/main/hooks/use-routes) which allows me to dynamically create and pre-filter routes, and render navigation from the definitions.

**Comment by brillout** [comment]
For folks who need/want Single Route File, please PM on discord.

**Comment by redbar0n** [comment]
FFR, here is a thread with people who don't like file based routing: https://twitter.com/t3dotgg/status/1583285667768827904

**Comment by CanRau** [comment]
I think SolidStart is taking a great approach with the dynamic `<FileRoutes />` component https://start.solidjs.com/core-concepts/routing

**Comment by redbar0n** [comment]
> I think SolidStart is taking a great approach with the dynamic `<FileRoutes />` component https://start.solidjs.com/core-concepts/routing

Yeah, I agree, with one exception. To quote it:

> Under the hood, SolidStart traverses your routes directory, collects all the routes, and makes them accessible using the [<FileRoutes />](https://start.solidjs.com/api/FileRoutes) component. The [<FileRoutes />](https://start.solidjs.com/api/FileRoutes) component only includes your UI routes, and not your API routes. You can use it instead of manually entering all your Routes inside the <Routes /> component in root.tsx. Let the compiler do the boring work!

I think it makes more sense to expose the file routes through some other means than as a UI component. Maybe as a function (e.g. import `getFileSystemRoutes()` like the `getRoute` that [next-type-safe-routes](https://github.com/ckastbjerg/next-type-safe-routes) does).

You then use it to import the default routes derived from the file system, which you are then able to overwrite in the routes file. Or you could choose to not import it and rather manually write out all the routes if you'd like. I think it could be the best of both worlds: convenience of filesystem routing (aka. file based routes) + the power and customizability of a single route file.

It would also allow vps to remain render library agnostic.

**Comment by brillout** [comment]
I'm actually thinking of increasingly focusing on single route files, and I'm increasingly thinking that it's superior than Next/Solid/Remix's approach. That said, so far, I think still supporting basic FS routing makes sense (like what VPS already does today), but I'll likely won't focus much on it.

**Comment by redbar0n** [comment]
[TanStack Router vs. React Router DOM vs. Next.js](https://tanstack.com/router/v1/docs/comparison) comparison page is out, as [per demand](https://github.com/TanStack/router/discussions/44). It has a good overview of potentially desirable features for a routing solution.

**Comment by brillout** [comment]
> [TanStack Router vs. React Router DOM vs. Next.js](https://tanstack.com/router/v1/docs/comparison) comparison page is out, as [per demand](https://github.com/TanStack/router/discussions/44). It has a good overview of potentially desirable features for a routing solution.

Other than typesafe links (which is quite lovely), is there something you'd miss if using VPS's built-in router?

**Comment by redbar0n** [comment]
Not that I can think of (the search params API seems neat, but not sure if it's in scope).

The ability to easily swap out the router and use any router would be nice. Even if it required a one-time mapping from the vps router to that router (say, for SSR).

**Comment by brillout** [comment]
> easily swap out the router

In principle, it should be possible to deeply integrate any router using the [`onBeforeRoute()` hook](https://vite-plugin-ssr.com/onBeforeRoute).

But, tbh and AFAICT, I see VPS's router to be superior for its intended purpose than what I'm seeing so far.

The only thing missing is typesafe links, which require either code gen or opting-out of Filesystem Routing (i.e. what this ticket is about). (AFAICT TypeScript types can't be deduced from the mere presence of a file, so it's either or.) I think it's possible to do code gen in a seamless manner without the user even noticing.

**Comment by brillout** [comment]
Ticket for typesafe links: https://github.com/brillout/vite-plugin-ssr/issues/698.

**Comment by brillout** [comment]
Now that the V1 design is out, this ticket should be fairly easy to implement.

To increase the priority of this ticket, add a comment here with the reasons why you want this.

**Comment by phiberber** [comment]
My approach to the Single Route File would be to use it into my framework, I'd like to have a File Routing System but I like it in my own style, that's why I would probably in a Single Route File declared in the renderer folder, use import.meta.glob or something like Chokidar. 

I like the idea of `+config` being the way to define routes, as I've seen in the documentation, you already can define the Page and if I remember well even onBeforeRender on the +config file, I'd think making it an object of paths would be an optimal way to make it.

```ts
// +config.js
export default {
  routes: {
    "/:id": {
     Page: UserPage,
     onBeforeRender: onBeforeRenderUser,

     "/activities": {
       Page: UserActivitiesPage,
       onBeforeRender: onBeforeRenderUserActivities
     }

     // ...etc
    }
  }
}
```

I'm not that sure about the naming of the property "routes"  as I don't know how it would be to separate API Routes and UI Routes.

**Comment by YannBirba** [comment]
Hello, 

I would be interested to get a single file routes definition to increase maintenability and also getting better git diff when updating routes.
I think fsr Is a bad idea as I said for git diff but also it forces the ecosystem to use specific file convention and it's not that flexible. 

It's also more user dev friendly to have a file to defined all routes, dev can quickly read it and understand what is going on at a specific url. I think vps is a plug and play ssr plugin for vite sooo imo it need to be as flexible as he can be.

I just want to be able to define my urls and get it strongly typed when I'm using a link. I also think child routes are not that good for readability I prefer to have a full relative url on each route.

Hope it can be helpful, sry for bad english. Also hope it's understandable.

**Comment by brillout** [comment]
>  it's not that flexible

Do you mind elaborating?

**Comment by rossanmol** [comment]
> 

A good example would be, where routes are managed externally from the code, in applications with multi project setup. Having the flexibility to define route pages without being forced to use file based routing, would be a great feature that Vite-Plugin-SSR could implement.

**Comment by brillout** [comment]
> Vite-Plugin-SSR could implement

Definitely and especially now with the V1 design [header files](https://vite-plugin-ssr.com/header-file).

> where routes are managed externally from the code, in applications with multi project setup.

Curious: what's your motivation for organizing your code like that?

The more I know about use cases, the better.

**Comment by rossanmol** [comment]
> > Vite-Plugin-SSR could implement
> 
> Definitely and especially now with the V1 design [header files](https://vite-plugin-ssr.com/header-file).
> 
> > where routes are managed externally from the code, in applications with multi project setup.
> 
> Curious: what's your motivation for organizing your code like that?
> 
> The more I know about use cases, the better.

It could be useful in projects where same codebase is used for multiple projects, and is managed externally via some kind of CMS. Pages, entire features are configurable without requiring a release.

**Comment by rossanmol** [comment]
> > Vite-Plugin-SSR could implement
> 
> Definitely and especially now with the V1 design [header files](https://vite-plugin-ssr.com/header-file).
> 
> > where routes are managed externally from the code, in applications with multi project setup.
> 
> Curious: what's your motivation for organizing your code like that?
> 
> The more I know about use cases, the better.

Do you know of any short term solution which could allow me to configure routes without file based routing?

**Comment by brillout** [comment]
> same codebase is used for multiple projects

Makes sense. What is it you're building? Let's PM. Let's also talk about sponsoring, which can make sense for your company as you'll be using a unique added value of VPS.

It's actually already implemented by [config.extends](https://vite-plugin-ssr.com/extends). But you can't currently provide pages (i.e. routes). This ticket is precisely about enabling defining pages/routes in [`+config.h.js`](https://vite-plugin-ssr.com/config) and therefore providing pages with `extends`.

For others reading this: if your company is up for sponsoring, definitely PM me. I prioritize features by the needs of sponsors.

**Comment by brillout** [comment]
I'm currently *not* considering implementing Single Route File for the https://github.com/brillout/vite-plugin-ssr/labels/v1%20release%20%3Astar2%3A.

AFAICT it's about two things:
 1. Being able to quickly glance over the entire routing of the app.
 2. Being able to define pages by external npm packages.

Are there other use cases? If so, let me know.

The thing is that I've a solution in mind that doesn't require Single Route File for both `1.` and `2.` and that's why, so far, I'm inclined to not implement this. But I'm happy to be convinced otherwise: let me know about your use case.

> The alternative solution for `1.` is a command `$ vite-plugin-ssr routes` that prints all routes. This command will automatically be run at the end of `$ vite build`.

**Comment by snake-py** [comment]
For me this would be an awesome feature for localization.

I could then dynamically change the routing based on the localization of the user. 

For instance:

in english this would be 

```
/home
/customer-data
```
in german
```
/home
/kunden-daten
```
 
It would make my life much easier to define this in a single file instead of in multiple files.

**Comment by snake-py** [comment]
@brillout my statement holds true though. I really would prefer to define my routes in a single file instead of having filesystem routing. If I can help to implement this let me know where I could get started.

**Comment by brillout** [comment]
A Single Route File also has its drawbacks. It isn't a perfect solution. I think I understand the appeal of it and I still have mixed feelings about it.

I'm thinking at this point the best is to first try to solve all issues without Single Route File and then later reconsider whether we want to add support for it.

That said, feel free to elaborate and why exactly you'd prefer a Single Route File over Filesystem Routing. (Your previous comment was more related to #1000. FYI I've a design in mind about i18n which I'll share if I'm still happy about it in a couple of days.)

**Comment by brillout** [comment]
That makes a lot of sense.

> I could have a single API

That's quite appealing indeed.

Actually, the V1 design needs a rudimentary implementation of a Single Route File. I'll see if I can get something useable for users.

**Comment by snake-py** [comment]
> I'll see if I can get something useable for users

@brillout if it helps you can also point me into a direction where and maybe "how" I could get started on, since I am need of the feature I would be willing to help out to maybe build a first implementation of this. I guess you have a lot of work to do with the V1 Design. However, I also know that first contributors can often cost more time than they bring to the table. 

So if contribution are welcome (and possible) I could have a look in coming up with an implementation if you have ideas for architecture maybe we can discuss them closer per pm (discord)? But I can also see that as maintainer you want to implemt this kind of core feature yourself.

**Comment by brillout** [comment]
I definitely appreciate the intention but, yeah, this one is quite a subtle one so I think it's best I implement it. (There are other tickets that are well suited for getting started with contributing: https://github.com/brillout/vite-plugin-ssr/issues?q=is%3Aissue+is%3Aopen+label%3A%22contribution-welcome+%3Atwo_hearts%3A%22 &mdash; in case any of it is appealing to you.)

**Comment by doeixd** [comment]
This would also be useful for any plugin that wants to build an abstraction above Vike/VPS . The Single Route File could be generated by them yet still be transparently visible/editable by the user. Be it an abstraction that provides the boilerplate for an API layer, internationalized routes, etc.

**Comment by brillout** [comment]
@doeixd You're right and that's the plan: https://vite-plugin-ssr.com/extends. (The rudimentary Single Route File implementation I mentioned earlier needed for the V1 design is actually precisely for enabling extensions to define pages.)

**Comment by brillout** [comment]
@redbar0n 

Each `+` file can live in a different environment which was a primary motivation for the V1 design; some of your ideas aren't implementable without code extraction (wich I believe to be an anti-pattern).

As for increased flexibility around Filesystem Routing I agree that'd be nice, but it isn't a priority at the moment. That said, I'm happy to revisit prioritisation upon blockers and/or sponsoring.

**Comment by brillout** [comment]
Latest design for defining all routes in a single file: https://github.com/vikejs/vike/issues/341#issue-1274652261.

**Comment by redbar0n** [comment]
> some of your ideas aren't implementable without code extraction

I presume you mean «[Code Extraction](https://www.builder.io/blog/wtf-is-code-extraction): The act of separating the code and creating server and client code bundles». But doesn’t Vike already do this to an extent?

If the problem with having all routes in a single file is that client side routing and server side routing would need to be extracted from that file, what about 2 files: a single route file for the client, and a single route file for the server? Could make it clearer. E.g. So if you want to SSR a particular page you’d move it to the single route file for the server.

**Comment by brillout** [comment]
@redbar0n I'm afraid it isn't that simple. In general, design needs an holistic understanding of a *lot* of problems at once. I'd be happy to elaborate but that would require 2-3 pages of text. As you can imagine, prioritization is crucial for Vike to grow.

That said, it's paramount that users communicate their pain points and what they try to achieve (the goals), and I'm more than happy to hear (critical) feedback here. The resulting design (how to achieve the goals) is a work that only Vike maintainers can do.

**Comment by brillout** [comment]
I think the crux of it is that Vike doesn't do code extraction, so I guess the usefulness of this is inherently limited.

To be clear, I ain't against this (and we'll actually implement something like this as some Vike extensions will need it), I'm just skeptical of its DX. It'd be nice to see how it works out in practice. I think TanStack goes a bit in that direction?
